File: Python3/0938. Range Sum of BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        queue = [root]
        res = 0
        while queue:
            node = queue.pop(0)
            if(node.val >= low and node.val <= high):
                res += node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            elif(node.val < low and node.right):
                queue.append(node.right)
            elif(node.val > high and node.left):
                queue.append(node.left)

        return res

File: C++/0938. Range Sum of BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        queue<TreeNode*> q;
        q.push(root);
        int res = 0;

        while(!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            if(node->val >= low && node->val <= high) {
                res += node->val;
                if(node->left != nullptr) {
                    q.push(node->left);
                }
                if(node->right != nullptr) {
                    q.push(node->right);
                }
            } else if (node->val < low && node->right != nullptr) {
                q.push(node->right);
            } else if (node->val > high && node->left != nullptr) {
                q.push(node->left);
            }
        }
        return res;

    }
};
